cogs/plugins.py

```diff
@@ -529,17 +541,16 @@ async def plugins_reset(self, ctx):
         for ext in list(self.bot.extensions):
             if not ext.startswith("plugins."):
                 continue
+            logger.error("Unloading plugin: %s.", ext)
             try:
-                logger.error("Unloading plugin: %s.", ext)
-                await self.bot.unload_extension(ext)
-            except Exception:
-                logger.error("Failed to unload plugin: %s.", ext)
-            else:
-                if not self.loaded_plugins:
-                    continue
                 plugin = next((p for p in self.loaded_plugins if p.ext_string == ext), None)
                 if plugin:
+                    await self.unload_plugin(plugin)
                     self.loaded_plugins.remove(plugin)
+                else:
+                    await self.bot.unload_extension(ext)
```

Comment:
`Bot.unload_extension` will handle it internally.

Actually I couldn't think of any case that would lead the code to reach here (an extension in the plugins dir but not in `loaded_plugins`). I just put it in there just in case tho, since we're purging the plugins here, so we would want to remove extensions found in their directory as well (handled with `Bot.unload_extension`), right?

It's how we used to do when purging plugins.

Reply:
@Jerrie-Aries If `unload_extension` does it internally, what's the point of the new `unload_plugin` function? `unload_plugin` calls `unload_extension()` and then deletes `sys.modules` that's under the parent path.

My original concern was, why doesn't extensions that are unloaded here gets the same treatment? And I'm guessing maybe a partially loaded plugin could reach this block (not sure). 

Perhaps a solution could be at the end of the ext loop in `plugins_reset`, we can iterate through `sys.modules` (as you've done in L274 above), and delete everything that starts the with plugin base path, to delete any leftovers.

Reply:
My apologies. I think I didn't get what you were asking initially. Yeah, I get it now.
The original purpose for `unload_plugin` method is to extend the `unload_extension` a little by removing modules loaded from the plugin's parent path. And currently it takes `Plugin` class as parameter, while the extension (partially loaded or broken plugin) in *that `else`* block is just a string. So I cannot use the `unload_plugin` there, or I couldn't think of a clean way to remove it 😔.
I guess your suggested solution could solve it once and for all to delete the leftovers.

So I may add after the `for ext ...` block:
```py
for module in list(sys.modules.keys()):
    if module.startswith("plugins."):
        del sys.modules[module]
```
This would do right?

Reply:
No worries at all. Yes I believe so, that would be my idea (not sure if it's good or bad). 
But if you have a better way of doing this, by all means feel free!

Reply:
Pushed a change as suggested. Please check.